### ISSUE TITLE
Application crashes in Debug mode and doesn't run as it should in Release mode when run from bin directories

### ISSUE THREAD

**Issue by HerrDingenz** [TheCherno/Hazel#199]
#### Describe the bug
As long as you build and run Hazel/Sandbox inside VS all is fine.
But when you run it from either Debug or Release build folders it either ASSERTS (Debug mode) or has no proper shader (Release mode).

#### To Reproduce
1. Build the Solution in Debug and Release modes.
2. Go to the respective bin folders and run the exe files there (either debug or release)
3. In Debug it simply ASSERTS being unable to load the shader files from the assets folder.
4. In Release mode it simply continues with the standard shader provided by the gfx card manufacturer (at least it does so here in my environment)

#### Expected behavior
It should work as intended ofcourse - load the shader file from the assets folder. But ofcourse it can't load it because the assets folder doesn't there (same issue we had with the hazel.dll at the beginning of our development journey before static linking).

#### Operating system: (please complete the following information)
- OS: Win64
- Version: Windows 7 Ultimate

#### Possible solution(s)
We could do a postbuild step copying the assets folder to the bin folder - which is pretty stupid if you ask me.
Or we simply do a postbuild step creating a link to the assets folder of the sandbox project using mklink (on windows).

In the project settings for sandbox in visual studio it would look like this:

mklink /J $(TargetDir)assets $(ProjectDir)assets

and it works perfectly fine.

I am just too stupid to put it inside the premake5.lua file :frowning_face:

**Comment by LovelySanta** [comment]
I was already aware of this, and we have a wobbly solution in the linux PR, but we keep going back to the postbuild step to copy assets to the bin folder as being the best solution. When you ship a game, the assets will also be next to the exe anyway, so I don't see why this is an issue... (except from the extra step)

**Comment by freeman40** [comment]
Even if you copy the "assets" directory to the bin folder then it will still crash if you cd one level up from bin, and then run  bin/Sandbox.exe.  (because now its looking for assets one level up from bin)   The issue is that the app is making a not-always-correct assumption about where the current working directory is.   Maybe at startup the app could store the directory part of "argv[0]" and then always load assets relative to that.

**Comment by LovelySanta** [comment]
Yes, this is something the System class should handle to obtain the correct path and then some FileSystem to use that info to make absolute paths. This was also suggested on Trello the other day. This is not something that I'll merge through PRs, but rather something Cherno has to implement during a video. However, for hazel as of now, this is quite far behind. I think at some point when linking assets through lua, this might come into play at some point, to resolve the file paths.

**Comment by IshoAntar** [comment]
Getting the application's root directory (the executable) is quite straight forward, actually.

Linux provides a [getcwd](https://linux.die.net/man/3/getcwd) function that returns the *current working directory* where the user launched the application from, together with `argv[0]` would give the root directory of Hazel.

for example, lets say the user has Hazel installed in `~/Hazel`,
executing it from the Home directory (i.e. `Hazel/bin/Hazel`) would result in:
`argv[0] == Hazel/bin/Hazel`
and
`getcwd == /home/<user>`

Concatenating them will result in an absolute path to the executable.

the code would look something like:
```
std::string hazelDir;
hazelDir.resize(255);
getcwd(&hazelDir[0], hazelDir.capacity());
hazelDir += '/';
hazelDir += argv[0];
```

Windows also provides a similar [_getcwd](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/getcwd-wgetcwd?view=vs-2019) function.

**Comment by LovelySanta** [comment]
That would indeed be a good solution, however implementing it is not. This would mean that we need some System class, which would also need to be platform dependent. I think this is something cherno would have to teach in his series first before merging.

**Comment by IshoAntar** [comment]
I agree that cherno should implement it in a video.

Also, after a quick [look around](https://stackoverflow.com/questions/2868680/what-is-a-cross-platform-way-to-get-the-current-directory#2868729), looks like `std::filesystem` has a `current_path()` function that does what `getcwd` does but in a more "standard" way, that could be considered when implementing this functionality.

**Comment by freeman40** [comment]
I don't think you need the absolute path, and you don't need to know what the current working directory is.  What you do need is the app path relative to current working directory (and then prepend this same relative path to any calls to load assets).  Given that Hazel provides the application entry point, I think the simplest solution is to just pass argv[0] from the entry point to the Application class.  Application stashes it away somewhere for later use (i.e. when loading assets).   You do also need the post-build step to copy the assets to app folder - I don't see this as "stupid", as LovelySanta says, this is what you would do when you ship the app anyway.
